src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -183,6 +122,131 @@ public GeneratorDriver AddAdditionalTexts(ImmutableArray<AdditionalText> additio
+                }
+
+                results.Add(new GeneratorRunResult(generator, sourceResults.ToImmutableAndFree(), generatorState.Diagnostics, generatorState.Exception));
+                trees.AddRange(generatorState.Trees);
+                diagnostics.AddRange(generatorState.Diagnostics);
+            }
+
+            return new GeneratorDriverRunResult(results.ToImmutableAndFree(), trees.ToImmutableAndFree(), diagnostics.ToImmutableAndFree());
+        }
+
+        internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, DiagnosticBag diagnosticsBag, CancellationToken cancellationToken = default)
+        {
+            // with no generators, there is no work to do
+            if (_state.Generators.IsEmpty)
+            {
+                return _state;
+            }
+
+            // run the actual generation
+            var state = StateWithPendingEditsApplied(_state);
+            var stateBuilder = ArrayBuilder<GeneratorState>.GetInstance();
+            int receiverCount = 0;
+
+            for (int i = 0; i < state.Generators.Length; i++)
+            {
+                var generator = state.Generators[i];
+                var generatorState = state.GeneratorStates[i];
+
+                // initialize the generator if needed
+                if (!generatorState.Info.Initialized)
+                {
+                    generatorState = InitializeGenerator(generator, diagnosticsBag, cancellationToken);
+                }
+
+                // create the syntax receiver if requested
+                if (generatorState.Info.SyntaxReceiverCreator is object)
+                {
+                    try
+                    {
+                        var rx = generatorState.Info.SyntaxReceiverCreator();
+                        generatorState = generatorState.WithReceiver(rx);
+                        receiverCount++;
+                    }
+                    catch (Exception e)
+                    {
+                        generatorState = SetGeneratorException(generatorState, generator, e, diagnosticsBag);
+                    }
+                }
+
+                stateBuilder.Add(generatorState);
+            }
+
+            // Run a syntax walk if any of the generators requested it
+            if (receiverCount > 0)
+            {
+                for (int i = 0; i < stateBuilder.Count; i++)
+                {
+                    var generatorState = stateBuilder[i];
+                    if (generatorState.SyntaxReceiver is object)
+                    {
+                        GeneratorSyntaxWalker walker = new GeneratorSyntaxWalker(generatorState.SyntaxReceiver);
+                        foreach (var tree in compilation.SyntaxTrees)
+                        {
+                            try
+                            {
+                                walker.Visit(tree.GetRoot(cancellationToken));
```

Comment:
>cancellationToken [](start = 58, length = 17)

If the containing task has been cancelled, will the containing `try / catch` suppress that? It seems like the call to `SyntaxTree.GetRoot()` should be outside the `try / catch` and any `OperationCanceledException` should be allowed to reach the caller. Is it possible to test that? Perhaps by having a generator `throw new OperationCanceledException()`?

Reply:
Ooh, great catch! I'll try and come up with a test to validate it.

---
In reply to: [464692102](https://github.com/dotnet/roslyn/pull/46296#discussion_r464692102) [](ancestors = 464692102)

Reply:
Hmm, I can't easily make a test for it, because none of the syntax trees in the compiler layer actually respect the `cancellationToken` parameter. I've made the change though because the workspace layer does, and it should prevent that from occuring. I added a test that check the same thing during execution doesn't cause it though which is useful and opened an [issue](https://github.com/dotnet/roslyn/issues/46528) to ensure we're properly handling the `cancellationToken` throughout the driver in general.

---
In reply to: [464717110](https://github.com/dotnet/roslyn/pull/46296#discussion_r464717110) [](ancestors = 464717110,464692102)